apollo-ios-codegen/Sources/GraphQLCompiler/JavaScript/src/utilities/fieldPolicyDirective.ts

```diff
@@ -0,0 +1,272 @@
+        { nodes: directive }
+      );
+    }
+
+    // check that the field exists
+    const actualField = typeFields[forField]
+    if (!actualField) {
+      throw new GraphQLError(
+        `Field "${forField}" does not exist on type "${type.name}".`,
+        { nodes: type.astNode ? [type.astNode, directive] : directive}
+      );
+    }
+
+    // validate the provided key args match an input parameter
+    validateKeyArgs(actualField, keyArgs);
+
+    // List input and return type validation
+    validateListRules(actualField);
+
+    (actualField as any)._apolloFieldPolicies = keyArgs;
+  }
+}
+
+function fieldPolicyDirectivesFor(
+  type: GraphQLCompositeType
+): DirectiveNode[] | undefined {
+  if(!isObjectType(type) && !isInterfaceType(type)) {
+    return undefined;
+  }
+
+  const result: DirectiveNode[] = [];
+
+  for (const extension of type.extensionASTNodes ?? []) {
+    const directive = extension.directives?.find(
+      (d) => d.name.value === directiveName
+    );
+    if (directive) {
+      result.push(directive)
+    }
+  }
+
+  for (const directive of type.astNode?.directives ?? []) {
+    if (directive.name.value === directiveName) {
+      result.push(directive);
+    }
+  }
+
+  if("getInterfaces" in type) {
+    for (const interfaceType of type.getInterfaces()) {
+      const found = fieldPolicyDirectivesFor(interfaceType);
+      if (!found) {
+        continue;
+      }
+
+      for (const foundDirective of found) {
+        var duplicate = false;
+        for (const directive of result) {
+          if (matchDirectiveArguments(directive, foundDirective)) {
+            duplicate = true;
+            break;
+          }
+        }
+
+        if (!duplicate) {
+          result.push(foundDirective);
+        }
+      }
+    }
+  }
+
+  return result
+}
+
+function matchDirectiveArguments(
+  first: DirectiveNode,
+  second: DirectiveNode
+): boolean {
+  return (
+    (first.arguments ?? [])
+      .map((node) =>
+        JSON.stringify([node.name.value, valueFromASTUntyped(node.value)])
+      )
+      .sort()
+      .toString() ===
+    (second.arguments ?? [])
+      .map((node) =>
+        JSON.stringify([node.name.value, valueFromASTUntyped(node.value)])
+      )
+      .sort()
+      .toString()
+  );
+}
+
+function validateKeyArgs(
+  actualField: GraphQLField<any, any>,
+  keyArgs: string[]
+): void {
+  for (const keyArg of keyArgs) {
+    validateKeyArgPath(actualField, keyArg);
+  }
+}
+
+function validateKeyArgPath(
+  field: GraphQLField<any, any>,
+  keyArg: string
+): void {
+  const parts = keyArg.split(".").filter(Boolean);
+  const [argName, ...path] = parts;
+
+  const arg = field.args.find(a => a.name === argName);
+  if (!arg) {
+    throw new GraphQLError(
+      `@fieldPolicy key argument "${keyArg}" does not exist as an input argument of field "${field.name}".`,
+      { nodes: field.astNode }
+    );
+  }
+
+  let curType: GraphQLInputType = base(arg.type);
```

Comment:
I'm assuming "cur" is short for current? I'd prefer just using "current" here. I'm fine with using universally commonly used abbreviations, like "arg" for "argument".